File: content/snippets/typescript/design_patterns/bridge.ts
interface Sender {
  send(message: string): void;
}

class EmailSender implements Sender {
  send(message: string): void {
    console.log(`email: ${message}`);
  }
}

class SmsSender implements Sender {
  send(message: string): void {
    console.log(`sms: ${message}`);
  }
}

class AlertNotification {
  constructor(private readonly sender: Sender) {}

  notify(message: string): void {
    this.sender.send(`alert -> ${message}`);
  }
}

new AlertNotification(new EmailSender()).notify('CPU high');
new AlertNotification(new SmsSender()).notify('CPU high');

File: content/snippets/go/design_patterns/bridge.go
package main

import "fmt"

type Sender interface {
	Send(message string)
}

type EmailSender struct{}

func (EmailSender) Send(message string) {
	fmt.Println("email:", message)
}

type SMSSender struct{}

func (SMSSender) Send(message string) {
	fmt.Println("sms:", message)
}

type AlertNotification struct {
	sender Sender
}

func (n AlertNotification) Notify(message string) {
	n.sender.Send("alert -> " + message)
}

func main() {
	AlertNotification{sender: EmailSender{}}.Notify("CPU high")
	AlertNotification{sender: SMSSender{}}.Notify("CPU high")
}
